src/shared/utilities/childProcess.ts

```diff
@@ -108,22 +127,85 @@ export class ChildProcess {
             this.childProcess!.stderr?.removeAllListeners()
             this.childProcess!.removeAllListeners()
         })
+
+        // Emitted when process exits.
+        // https://nodejs.org/api/child_process.html#child_process_class_childprocess
+        this.childProcess.once('exit', (code, signal) => {
+            this.processStopped = true
+        })
+    }
+
+    /**
+     * Gets the `run()` result after the child process has finished.
+     *
+     * @returns `run()` result, or undefined if the process has not yet started or is still running.
+     */
+    public result(): ChildProcessResult | undefined {
+        return this.processResult
+    }
+
+    public pid(): number {
+        return this.childProcess?.pid ?? -1
     }
 
-    public kill(): void {
-        if (this.childProcess && !this.killed) {
-            this.childProcess.kill()
+    public exitCode(): number {
+        return typeof this.childProcess?.exitCode == 'number' ? this.childProcess.exitCode : -1
+    }
+
+    /**
+     * Stops the process.
+     *
+     * SIGTERM won't kill a terminal process, use SIGHUP instead.
+     *
+     * @param force  Tries SIGKILL if the process is not stopped after a few seconds.
+     * @param signal  Signal to send, defaults to SIGTERM.
+     *
+     */
+    public stop(force?: boolean, signal?: string): void {
+        const child = this.childProcess
+        if (!child) {
+            return
+        }
+        const command = this.command
+        const pid = this.pid()
+        if (!this.stopped) {
+            child.kill(signal)
+
+            if (force === true) {
+                waitUntil(
+                    async () => {
+                        return this.stopped
+                    },
+                    { timeout: 3000, interval: 200 }
+                )
+                    .then(stopped => {
+                        if (!stopped) {
+                            child.kill('SIGKILL')
+                        }
+                    })
+                    .catch(e => {
+                        getLogger().warn(`stop(): SIGKILL failed: pid=${pid} command=${command}`)
+                    })
+            }
         } else {
             throw new Error('Attempting to kill a process that has already been killed')
         }
     }
 
-    public get killed(): boolean {
-        // default to true for safety
-        return this.childProcess ? this.childProcess.killed : true
+    /**
+     * Returns true if the process has ended (streams are closed and/or exitCode was set).
+     *
+     * Returns false if the process was not started or is still running.
+     */
+    public get stopped(): boolean {
```

Comment:
What's the distinction for this over checking whether or not `this.result` is falsy? Is there a point where `this.result` is defined but this isn't?

Reply:
great question! I think you're right,  `this.childProcess && this.processResult === undefined` tells us "process was started and not yet stopped". So I will make that change.

Reply:
@bryceitoc9 Latest push implements that; it required handling onExit because 'close' event may not fire in some cases (presumably when the process is stopped immediately after starting, so streams aren't set up?).

Updated PR description to mention these changes:

> - Introduce `collect` option which controls whether `start()` collects stdout/stderr (this allows `stopped()` to check `this.processResult` to decide if the process stopped).
> - Move stdout/stderr collection to `start()`, so `run()` is now just a thin wrapper.